Core/clim-basic/setf-star.lisp

```diff
@@ -33,37 +33,42 @@
 			    (symbol-name '#:star))
 		    (symbol-package name-sym)))))
 
-(defun make-arglist (args)
-  "Convert the list ARGS to a list of arguments in the form (var
-init-form supplied-p-parameter)."
-  (mapcar (lambda (a) `(,a nil ,(gensym))) args))
-
 (defmacro defgeneric* (fun-name lambda-list &body options)
   "Defines a SETF* generic function.  FUN-NAME is a SETF function
 name.  The last required argument is the single argument to the function in a
 SETF place form; the other arguments are values collected from the
 SETF new value form."
   (unless (setf-name-p fun-name)
     (error "~S is not a valid name for a SETF* generic function." fun-name))
-  (let ((setf-name (cadr fun-name))
-        (gf (make-setf*-gfn-name fun-name)))
-    (multiple-value-bind (required optional rest keys allow-other-keys aux keyp)
-        (alexandria:parse-ordinary-lambda-list lambda-list)
-      (when aux
-        (error "The use of &aux is not allowed in generic function lambda list."))
-      (let ((optional (make-arglist (mapcar #'car optional)))
-            (keys (make-arglist (mapcar #'cadar keys))))
+  (multiple-value-bind (required optional rest keys allow-other-keys aux keyp)
+      (alexandria:parse-ordinary-lambda-list lambda-list)
+    (when aux
+      (error "The use of &aux is not allowed in generic function lambda list."))
+    (let ((setf-name (cadr fun-name))
+          (gf (make-setf*-gfn-name fun-name))
+          (args (butlast required))
+          (place (alexandria:lastcar required))
+          (optional (mapcar (lambda (opt)
+                              `(,(first opt) nil ,(gensym)))
+                            optional))
+          (key-args (mapcar (lambda (key)
+                              (destructuring-bind ((key arg) val suppliedp) key
+                                (declare (ignore val suppliedp))
+                                `((,key ,arg) nil ,(gensym))))
+                            keys)))
+      (let ((call-opt (mapcar #'first optional))
+            (call-key (loop for ((key arg) default suppliedp) in key-args
+                            collect `(and ,suppliedp `(list ,,key ,,arg)))))
         `(progn
            (defsetf ,setf-name
-               (,(car (last required))
+               (,place
                 ,@(and optional `(&optional ,@optional))
                 ,@(and rest `(&rest ,rest))
-                ,@(and keyp `(&key ,@keys))
+                ,@(and keyp `(&key ,@key-args))
                 ,@(and allow-other-keys `(&allow-other-keys)))
-               ,(butlast required)
-             `(funcall #',',gf ,,@required
-                ,,@(mapcar (lambda (x) `(and ,(third x) ,(car x))) optional)
-                ,@,@(mapcar (lambda (x) `(and ,(third x) `(,,(make-keyword (car x)) ,,(car x)))) keys)))
+               ,args
+             `(apply #',',gf ,,@args ,,place
+                (append (list ,,@call-opt) ,,@call-key)))
```

Comment:
It would be nice to avoid the `apply`, `append`, `list` at runtime if `call-opt` and/or `call-key` are empty.

Reply:
see a new version of the code. it avoid apply and avoids appending at runtime (and list).